File: python/spinlab/spinrec.py
"""Read/write .spinrec binary input recording format."""
from __future__ import annotations

import struct
from dataclasses import dataclass

MAGIC = b"SREC"
VERSION = 1
HEADER_SIZE = 32
HEADER_FMT = "<4sH16sI6s"


@dataclass
class SpinrecHeader:
    magic: bytes
    version: int
    game_id: str
    frame_count: int


def read_spinrec(data: bytes) -> tuple[SpinrecHeader, list[int]]:
    """Parse a .spinrec binary blob into header + frame list."""
    if len(data) < HEADER_SIZE:
        raise ValueError(f"Too short for header: {len(data)} bytes (need {HEADER_SIZE})")
    magic, version, game_id_bytes, frame_count, _ = struct.unpack_from(HEADER_FMT, data)
    if magic != MAGIC:
        raise ValueError(f"Bad magic: {magic!r} (expected {MAGIC!r})")
    game_id = game_id_bytes.rstrip(b"\x00").decode("ascii")
    expected_body = frame_count * 2
    actual_body = len(data) - HEADER_SIZE
    if actual_body < expected_body:
        raise ValueError(f"Body truncated: {actual_body} bytes (expected {expected_body})")
    frames = list(struct.unpack_from(f"<{frame_count}H", data, HEADER_SIZE))
    return SpinrecHeader(magic=magic, version=version, game_id=game_id, frame_count=frame_count), frames


def write_spinrec(game_id: str, frames: list[int]) -> bytes:
    """Build a .spinrec binary blob."""
    header = struct.pack(HEADER_FMT, MAGIC, VERSION, game_id.encode("ascii"), len(frames), b"\x00" * 6)
    body = struct.pack(f"<{len(frames)}H", *frames) if frames else b""
    return header + body

File: lua/spinrec.lua
-- spinrec.lua — .spinrec binary format I/O and SNES input encoding.
-- All functions are global (loaded via dofile).
-- Depends on global log() from spinlab.lua.
--
-- .spinrec format: 32-byte header + one uint16/frame (SNES joypad bitmask).
-- Header: SREC (4) + version (2) + game_id (16) + frame_count (4) + reserved (6)

function flush_spinrec(path, game_id_str, buffer)
  local f = io.open(path, "wb")
  if not f then
    log("ERROR: Cannot write spinrec: " .. path)
    return false
  end
  -- Magic
  f:write("SREC")
  -- Version (uint16 LE)
  f:write(string.char(1, 0))
  -- Game ID (16 bytes ASCII, pad with zeros if shorter)
  local gid = (game_id_str or ""):sub(1, 16)
  f:write(gid .. string.rep("\0", 16 - #gid))
  -- Frame count (uint32 LE)
  local n = #buffer
  f:write(string.char(n & 0xFF, (n >> 8) & 0xFF, (n >> 16) & 0xFF, (n >> 24) & 0xFF))
  -- Reserved (6 zeros)
  f:write(string.rep("\0", 6))
  -- Body: 2 bytes per frame (uint16 LE)
  for _, mask in ipairs(buffer) do
    f:write(string.char(mask & 0xFF, (mask >> 8) & 0xFF))
  end
  f:close()
  log("Wrote spinrec: " .. path .. " (" .. n .. " frames)")
  return true
end

function read_spinrec(path)
  local f = io.open(path, "rb")
  if not f then return nil, "file not found: " .. path end
  local data = f:read("*a")
  f:close()
  if #data < 32 then return nil, "file too short" end
  -- Validate magic
  if data:sub(1, 4) ~= "SREC" then return nil, "bad magic" end
  -- Parse header
  local b = function(i) return data:byte(i) end
  local frame_count = b(23) + b(24) * 256 + b(25) * 65536 + b(26) * 16777216
  local gid = data:sub(7, 22)
  -- Validate body length
  local expected_body = frame_count * 2
  if #data - 32 < expected_body then return nil, "body truncated" end
  -- Parse frames
  local frames = {}
  for i = 1, frame_count do
    local offset = 32 + (i - 1) * 2
    frames[i] = b(offset + 1) + b(offset + 2) * 256
  end
  return {game_id = gid, frame_count = frame_count, frames = frames}, nil
end

-- Input bitmask encoding: matches SNES joypad register layout
local INPUT_BITS = {
  b = 0, y = 1, select = 2, start = 3,
  up = 4, down = 5, left = 6, right = 7,
  a = 8, x = 9, l = 10, r = 11,
}

function encode_input(tbl)
  local mask = 0
  for name, bit in pairs(INPUT_BITS) do
    if tbl[name] then mask = mask + (1 << bit) end
  end
  return mask
end

function decode_input(mask)
  local tbl = {}
  for name, bit in pairs(INPUT_BITS) do
    tbl[name] = (mask & (1 << bit)) ~= 0
  end
  return tbl
end
